### ISSUE TITLE
Merge all ecosystem builders into this repo

### ISSUE THREAD

**Issue by laurentsimon** [slsa-framework/slsa-github-generator#36]
I was discussing with @MarkLodato about the cost of maintenance of the (possibly) many builder repos we will eventually have. 

The main reason we opted to have separate ecosystem repos was to avoid asking users to update their workflow hash/version when another ecosystem's builder was updated. However, we're starting to wonder if this is the right thing to do or not.

An alternative would be to have all builders in this repository, which *could* reduce maintenance for us. I imagine it would look like this:
- Ecosystem builders: .github/workflows/slsa3_go.yml, .github/workflows/slsa3_ko.yml, etc
- Each builder's specific code could live under ./go, ./ko, etc. Maybe overtime we would find better ways to share code between them, e.g. thru a standard codebase with "interface/template" for each ecosystem: dry-run command, download dependency command, compilation command, etc. which would allow us to abstract away the underlying compiler

Thoughts?

cc @asraa @ianlewis @joshuagl

**Comment by joshuagl** [comment]
I suspect it will be easier to unbundle the ecosystem builders into separate repos at a later date, rather than bundle them into a consolidated repo once they become established, so I am in favour of having all builders in this repo in the near-term and trying to identify ways to code share.

+1 from me.

**Comment by ianlewis** [comment]
Yeah, I think it will be simpler to consolidate docs etc. so there's only one place to look for info/references as well and will make it easier to consume if we have a single repo.

Docs could be something like:
We provide the following workflows:
- Generic provenance-only
- Go
- Dockerfile/Docker image
- etc.

Here's how to create your own ecosystem builder:
...

**Comment by ianlewis** [comment]
Would it also make sense to include the verifier here as well? Might make it easier to manage potential breaking changes if they are both in one repo with one set of pre-submits etc.

**Comment by MarkLodato** [comment]
> Would it also make sense to include the verifier here as well? Might make it easier to manage potential breaking changes if they are both in one repo with one set of pre-submits etc.

My inclination is no, that is a different use case with different code. Plus, the verifier will hopefully support more than thus this generator in the future.

**Comment by laurentsimon** [comment]
I'm fine with that too. If we were to merge the verifier into this repo, I suppose we'd rename this repo to `slsa-tools` and have multiple commands like: `verify`, `generate`, etc.

**Comment by laurentsimon** [comment]
Assuming a single repo *without* the verifier, we'll have the generic API at the root folder and the ecosystem-specific ones, like `./go/ with their independent `./go/go.mod` and `./go/go.sum`, correct? 

Or you had something else in mind?

**Comment by ianlewis** [comment]
yeah, I think the generic API can be at the root, and the others in maybe a subdirectory like `ecosystem/go`, `ecosystem/ko` etc.  Something like that.

I've not played around with the new workspace feature in Go (i.e. where you can have multiple go.mod packages that can import from each other). I'd have to play around with it a bit before I could make a recommendation, but I'd imagine that maybe we could use that?

**Comment by laurentsimon** [comment]
> I've not played around with the new workspace feature in Go (i.e. where you can have multiple go.mod packages that can import from each other). I'd have to play around with it a bit before I could make a recommendation, but I'd imagine that maybe we could use that?

sure. So I will start migrating the existing go/ko code in here, and we can figure this importing feature later. I'm hoping we should be able to import the top-folder project as we currently do from other repos

**Comment by laurentsimon** [comment]
Once https://github.com/slsa-framework/slsa-github-generator-go/pull/125 is merged, I'll move the content of the https://github.com/slsa-framework/slsa-github-generator-go repo in this repo

**Comment by ianlewis** [comment]
> Once [slsa-framework/slsa-github-generator-go#125](https://github.com/slsa-framework/slsa-github-generator-go/pull/125) is merged, I'll move the content of the https://github.com/slsa-framework/slsa-github-generator-go repo in this repo

Once #125 is merged, do you want to migrate the issues over too?
https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository

**Comment by ianlewis** [comment]
@laurentsimon Also, do you want to include the https://github.com/laurentsimon/slsa-github-generator-ko eventually?

**Comment by laurentsimon** [comment]
This is done. I'll create a separate issue for ko builder
